### PR TITLE
Mirage3

### PR DESCRIPTION
**Pull request by samoht**
This is built on top of the major cleanup that I've pushed in #259

Everything compiles but it seems that there are 3 missing functions in the tcpip stack:

- `Stack_ipv4.set_ip_netmask`
- `Stack_ipv4.set_ip_gateways`
- `Stack_tcp_wire.src_of_id` (e.g. there is no way to get the source IP)

Also the tests are compiling but not passing, either because of my very flaky connection or because I broke something with the `dns` ref that I modified to be `None` by default. I will investigate more next week.

### CODE REVIEW

**Comment by djs55** [comment]
```
Error: External library "tcpip.unix" not found.
-> required by "src/hostnet/jbuild (context default)"
Hint: try: jbuilder external-lib-deps --missing --dev src/bin/main.exe
make: *** [vpnkit.exe] Error 1
```

**Comment by samoht** [comment]
It needs the dev version of `mirage-tcpip` and `charrua-core`. I'll fix this.

**Comment by samoht** [comment]
Hum, it seems that some of the tests in mirage-tcpip do not work on windows:

```
--------------------------------------------------------------------------------
- ASSERT sender address
- --------------------------------------------------------------------------------
- Fatal error: exception (Failure "Error sender address: expecting fc00::23, got ::.")
```

And the vpnkit tests are failing on macOS (I can reproduce this locally, and @djs55 identified the error message to come from [here](https://github.com/mirage/mirage-tcpip/blob/e7a9745256cf10366b06c4472a416068185761cd/src/ipv4/static_ipv4.ml#L57)):
```
main_uwt.exe: [WARNING] IP module couldn't send UDP packet to 8.8.8.8: no route to destination: no response for IP on local network
Fatal error: exception Dns__Protocol.Dns_resolve_error(_)
Raised at file "src/core/lwt.ml", line 2719, characters 24-25
Called from file "uwt.ml", line 1848, characters 10-23
Called from file "list.ml", line 77, characters 12-15
Called from file "list.ml", line 77, characters 12-15
Called from file "src/hostnet_test/main_uwt.ml", line 19, characters 2-282
make: *** [test] Error 2
```

**Comment by djs55** [comment]
I'll try to fix the routing problem by switching to DHCP in the test case. This requires: https://github.com/mirage/charrua-core/pull/65

**Comment by samoht** [comment]
I can produce a different failure for `ipv6` on Windows. Trying to understand what happens ...

**Comment by samoht** [comment]
There are 2 failing tests on windows:
- ipv6 0:
```
test.exe: [INFO] Connected Ethernet interface 02:50:00:00:00:01
test.exe: [INFO] IP6: Starting
test.exe: [DEBUG] ND6: Sending RS
test.exe: [DEBUG] ND6: Sending NS src=:: dst=ff02::1:ff00:1 tgt=fe80::50:ff:fe00:1
test.exe: [DEBUG] no size limit, sending
test.exe: [DEBUG] no size limit, sending
test.exe: [DEBUG] ND6: Sending NS src=:: dst=ff02::1:ff00:23 tgt=fc00::23
test.exe: [DEBUG] no size limit, sending
test.exe: [INFO] UDP interface connected on
test.exe: [INFO] Connected Ethernet interface 02:50:00:00:00:02
test.exe: [INFO] IP6: Starting
test.exe: [DEBUG] ND6: Sending RS
test.exe: [DEBUG] ND6: Sending NS src=:: dst=ff02::1:ff00:2 tgt=fe80::50:ff:fe00:2
test.exe: [DEBUG] no size limit, sending
test.exe: [DEBUG] no size limit, sending
test.exe: [DEBUG] ND6: Sending NS src=:: dst=ff02::1:ff00:45 tgt=fc00::45
test.exe: [DEBUG] no size limit, sending
test.exe: [INFO] UDP interface connected on
test.exe: [DEBUG] IP6: Queueing packet: dst=fc00::45
test.exe: [DEBUG] ND6: Sending NS src=:: dst=ff02::1:ff00:45 tgt=fc00::45
test.exe: [DEBUG] no size limit, sending
test.exe: [DEBUG] ND: Received NS: src=:: dst=ff02::1:ff00:45 tgt=fc00::45
test.exe: [DEBUG] SLAAC: fe80::50:ff:fe00:2 --> PREFERRED
test.exe: [DEBUG] SLAAC: fc00::45 --> PREFERRED
test.exe: [DEBUG] SLAAC: fe80::50:ff:fe00:1 --> PREFERRED
test.exe: [DEBUG] SLAAC: fc00::23 --> PREFERRED
--------------------------------------------------------------------------------------------------------------------
ASSERT UDP packet should have been received
--------------------------------------------------------------------------------------------------------------------
Test error: Error UDP packet should have been received.
```

- socket 1: 
```
_build/_tests\socket.001.output:
test.exe: [INFO] Manager: connect
test.exe: [INFO] Manager: configuring
test.exe: [WARNING] Manager: sockets currently bind to all available IPs. IPs 127.0.0.1 were specified, but this will be ignored
test.exe: [INFO] Manager: connect
test.exe: [INFO] Manager: configuring
test.exe: [WARNING] Manager: sockets currently bind to all available IPs. IPs 127.0.0.1 were specified, but this will be ignored
[exception] Unix.Unix_error(Unix.EPROTONOSUPPORT, "socket", "")
```

I will follow up on mirage/mirage-tcpip

**Comment by djs55** [comment]
One of the `tcpip` tests just failed on CircleCI too:
```
[ERROR]             connect              1   connect two stacks, uniform packet loss of packets with no payload x 100.
```
(probably intermittent)

**Comment by samoht** [comment]
yes sorry I have changed the default test flags in my tcpip branch, I will create a more stable dev branch ...

**Comment by samoht** [comment]
I've pushed a new Git url for tcpip dev package, where only quick tests are running. Not sure why the slow test is failing, but this is probably due to a timeout/slow machine.

File: src/hostnet/mux.ml [hunk 1]
```diff
@@ -91,9 +91,15 @@ module Make (Netif: Mirage_net_lwt.S) = struct
     let rules = RuleMap.empty in
     let default_callback = fun _ -> Lwt.return_unit in
     let t = { netif; rules; default_callback } in
-    Netif.listen netif @@ callback t >>= fun r ->
-    let r = match r with  Ok () -> Ok t | Error _ as e -> e in
-    Lwt.return r >|= lift_error
+    Lwt.async
+      (fun () ->
+        Netif.listen netif @@ callback t >>= function
```
Comment (samoht) [inline src/hostnet/mux.ml]: you can write this as:
```ocaml
 Netif.listen netif @@ callback t >|= function
| Ok ()   -> ()
| Error _ -> Log.err (fun f -> ... )
```

File: src/hostnet/slirp.ml [hunk 2]
```diff
@@ -269,17 +269,15 @@ module Filteredif = Filter.Make(Vmnet)
       Stack_ethif.connect ~mtu netif >>= fun ethif ->
       Stack_arpv4.connect ~table:arp_table ethif |>fun arp ->
       Stack_ipv4.connect
+        ~ip
         ~gateway:None
-        ~network:Ipaddr.V4.(Prefix.of_addr unspecified)
+        ~network:Ipaddr.V4.Prefix.global
```
Comment (samoht) [inline src/hostnet/slirp.ml]: good catch! I wasn't really sure what to use there :-)

**Comment by samoht** [comment]
@djs55 it now seems that we have genuine test errors on both windows and macOS:

```
**** Starting test Host resolver: lookup 
Starting test case 
main_uwt.exe: [INFO] Made a loopback connection
main_uwt.exe: [INFO] PPP.negotiate: received { magic = VMN3T; version = 1; commit = 0123456789012345678901234567890123456789 }
main_uwt.exe: [INFO] Client.negotiate: received { magic = VMN3T; version = 1; commit = 0123456789012345678901234567890123456789 }
main_uwt.exe: [INFO] PPP.negotiate: received Ethernet d1d9cd61-d0dc-4715-9bb3-4c11da7ad7a5
main_uwt.exe: [INFO] PPP.negotiate: sending { mtu = 1500; max_packet_size = 1550; client_macaddr = c0:ff:ee:c0:ff:ee }
main_uwt.exe: [INFO] PPP.listen: rebinding the primary listen callback
main_uwt.exe: [INFO] Client mac: c0:ff:ee:c0:ff:ee server mac: f6:16:36:bc:f9:c6
main_uwt.exe: [INFO] TCP/IP ready
main_uwt.exe: [INFO] stack connected
main_uwt.exe: [INFO] Initialising client TCP/IP stack
main_uwt.exe: [INFO] PPP.listen: rebinding the primary listen callback
main_uwt.exe: [ERROR] Socket.TCPv4.read tcp:127.0.0.1:49510: caught Lwt.Completion_loop.Canceled returning Eof
Fatal error: exception (Failure timeout)
Raised at file "src/core/lwt.ml", line 2719, characters 24-25
Called from file "uwt.ml", line 1848, characters 10-23
Called from file "list.ml", line 77, characters 12-15
Called from file "list.ml", line 77, characters 12-15
Called from file "src/hostnet_test/main_uwt.ml", line 19, characters 2-282
```

**Comment by djs55** [comment]
Tests are still failing in CI but I've not managed to reproduce it locally yet. I've pushed some patches to improve the debug logging to see if that helps. I'll see if I can make myself a more exact replica of the CI environment.

**Comment by djs55** [comment]
OK it reproduced for me when I ran `scripts/common.sh`. A bit of environment-bisecting later and the culpit appears to be: https://github.com/mirage/charrua-core/pull/66 -- the tests work for me when this is merged.

**Comment by samoht** [comment]
The new errors are related to missing license files, which means the tests passed, yay!!

I'll fix the licenses :-)

**Comment by samoht** [comment]
ALL GREEN!

The last worrying (or not) bit are the failing TCPIP tests on windows. Probably not really severe, but would be great to check that we understand what happens. I'll have a loop tomorrow if I find some time.

**Comment by djs55** [comment]
Was it only the IPv6 tests? If so that's not too bad, as we don't use IPv6 (yet).

Since I'm paranoid, I'll put this `vpnkit` build through the Docker for Desktop end-to-end tests too and report back.

**Comment by samoht** [comment]
yea, the ipv6 test is failing but also using an unsuported protocol over a socket: https://github.com/mirage/mirage-tcpip/issues/331

**Comment by djs55** [comment]
Docker Windows tests all passed but Mac tests all failed -- I'll investigate tomorrow. It's probably a startup issue somehow related to the `Vmnet` Unix domain socket.

**Comment by djs55** [comment]
@samoht sorry I force pushed this and lost 2 of your patches. I've re-applied them, hopefully it's ok now.

**Comment by djs55** [comment]
OK, the Docker Mac tests are passing now. I think we are safe to merge this. Any objections? (@samoht?)

**Pull request merged**